File: Week 03/id_608/search-a-2d-matrix.go
package main

import "fmt"

func main() {
	matrix := [][]int{
		{1, 3, 5, 7},
		{10, 11, 16, 20},
		{23, 30, 34, 50},
	}
	result := searchMatrix(matrix, 11)
	fmt.Println(result)
}

func searchMatrix(matrix [][]int, target int) bool {
	for _, value := range matrix {
		l := len(value)
		if l == 0 {
			return false
		}
		if target >= value[0] && target <= value[l-1] {
			left := 0
			right := l - 1
			for left <= right {
				mid := (left + right) / 2
				if target == value[mid] {
					return true
				} else {
					if target < value[mid] {
						right = mid - 1
					} else {
						left = mid + 1
					}
				}
			}
		}
	}
	return false
}

File: Week 03/id_328/SearchA2dMatrix.java
public class SearchA2dMatrix {
    public boolean searchMatrix(int[][] matrix, int target) {
        int rowCount = matrix.length;
        if(rowCount == 0 || matrix[0].length == 0) return false;
        int top = 0 ;
        int bottom = rowCount - 1;
        int mid = 0 ;
        if(rowCount > 1){
            while( top < bottom){
                mid = (top + bottom) /2;
                if(matrix[mid][0] == target){
                 return true;
                }else if(matrix[mid][0] < target){
                    top = mid + 1;
                }else {
                    bottom = mid - 1;
                }
            }
        
            if(matrix[top][0] > target ) top--;

            if(top < 0){
                return false;
            }

        }
        int left = 0 ;
        int right = matrix[top].length - 1;
        mid = 0;
        while(left <= right){
            mid = (left + right) / 2;
            if(matrix[top][mid] == target){
                return true;
            }else if(matrix[top][mid] < target){
                left = mid + 1;
            }else {
                right = mid - 1;
            }
        }
        return false;
    }
}